### PR TITLE
feat(maas): wire maxExpirationDays to ConfigMap with automatic pod restart

### PR DESCRIPTION
**Pull request by ishitasequeira**
## Description

This PR implements two related features for ModelsAsService:

### 1. Wire `apiKeys.maxExpirationDays` to ConfigMap
Connects the `spec.apiKeys.maxExpirationDays` field from the ModelsAsService CR to the maas-api deployment. When set, the operator updates the `maas-parameters` ConfigMap with the configured value, which is then consumed by the maas-api pod via environment variable.

### 2. Automatic Pod Restart on Config Changes
Adds a config hash annotation (`app.opendatahub.io/MaaSConfigHash`) to the maas-api Deployment's pod template. This hash is computed from the ConfigMap data, ensuring that any configuration change (including `maxExpirationDays`) automatically triggers a rolling restart of the maas-api pods.

**Why is this needed?**
Kubernetes does not automatically restart pods when ConfigMap data changes if the values are mounted via `valueFrom.configMapKeyRef`. The config hash annotation pattern ensures configuration changes are applied without manual intervention.

This builds on #481 which added the `apiKeys` field to the ModelsAsService CRD.

https://issues.redhat.com/browse/RHOAIENG-51784

### Unit Tests
- Tests for API key configuration (various values, nil cases, edge cases)
- Tests for config hash annotation (hash computation, update on change, missing resources)

### Manual Cluster Validation
1. Enabled MaaS in DataScienceCluster
2. Set `spec.apiKeys.maxExpirationDays: 90` on ModelsAsService CR
3. Verified ConfigMap `maas-parameters` contains `api-key-max-expiration-days: 90`
4. Changed value to `365` and verified:
   - ConfigMap updated automatically
   - Hash annotation changed from `7fa36...` to `18b35...`
   - New pod created (`maas-api-78b686dc8d-x4kpk`) replacing old pod
   - Rolling restart completed successfully

## Merge criteria

- [x] You have read the [contributors guide](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/CONTRIBUTING.md).
- [x] Commit messages are meaningful - have a clear and concise summary and detailed explanation of what was changed and why.
- [x] Pull Request contains a description of the solution, a link to the JIRA issue, and to any dependent or related Pull Request.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work
- [x] The developer has run the integration test pipeline and verified that it passed successfully

### E2E test suite update requirement

When bringing new changes to the operator code, such changes are by default required to be accompanied by extending and/or updating the E2E test suite accordingly.

To opt-out of this requirement:
1. **Please inspect the [opt-out guidelines](https://github.com/opendatahub-io/opendatahub-operator/blob/main/docs/e2e-update-requirement-guidelines.md)**, to determine if the nature of the PR changes allows for skipping this requirement
2. If opt-out is applicable, provide justification in the dedicated `E2E update requirement opt-out justification` section below
3. Check the checkbox below:
- [x] Skip requirement to update E2E test suite for this PR
4. Submit/save these changes to the PR description. This will automatically trigger the check.

#### E2E update requirement opt-out justification
The MaaS e2e tests are currently skipped in the operator test suite. The config hash pattern is an internal implementation detail that doesn't require dedicated e2e coverage - unit tests provide sufficient coverage for the hash computation and annotation logic.

## Summary by CodeRabbit

* **New Features**
  * Optional API key maximum-expiration-days can be set and is applied to API key parameters.
  * Automatic config-hash annotation on the MaaS service deployment so changes to MaaS parameters trigger a rolling restart.

* **Tests**
  * Added comprehensive tests for API key configuration (multiple values and edge cases).
  * Added tests verifying config-hash annotation behavior, deterministic hash output, and restart triggering when config changes.

### CODE REVIEW

File: internal/controller/components/modelsasservice/modelsasservice_controller_actions.go [hunk 1]
```diff
@@ -201,3 +218,113 @@ func configureDestinationRule(log logr.Logger, resource *unstructured.Unstructur
 
 	resource.SetNamespace(gatewayNamespace)
 }
+
+// configureConfigHashAnnotation adds a hash annotation to the maas-api Deployment
+// to trigger rolling restarts when the ConfigMap changes.
+// This is necessary because env vars sourced via valueFrom.configMapKeyRef
+// do not automatically trigger pod restarts when the ConfigMap is updated.
+func configureConfigHashAnnotation(ctx context.Context, rr *types.ReconciliationRequest) error {
```
Comment (bartoszmajsak) [inline internal/controller/components/modelsasservice/modelsasservice_controller_actions.go:226]: Why not just changing env var value directly instead? Wouldn't this trigger pod restart?

Comment (bartoszmajsak) [reply internal/controller/components/modelsasservice/modelsasservice_controller_actions.go:226]: I see now, the rest is [sourced the same way](https://github.com/opendatahub-io/models-as-a-service/blob/c4af5aae67237e5a2e7bfa020a241f6bd0d2963b/deployment/overlays/odh/kustomization.yaml#L23-L44). Fish memory. In such a case, this is fine.

File: internal/controller/components/modelsasservice/modelsasservice_controller_actions.go [hunk 2]
```diff
@@ -201,3 +218,113 @@ func configureDestinationRule(log logr.Logger, resource *unstructured.Unstructur
...
+		return nil
+	}
+
+	// Initialize annotations map if nil
+	if deployment.Spec.Template.Annotations == nil {
+		deployment.Spec.Template.Annotations = make(map[string]string)
+	}
+
+	// Add the config hash annotation to the pod template
+	annotationKey := labels.ODHAppPrefix + "/MaaSConfigHash"
```
Comment (bartoszmajsak) [inline internal/controller/components/modelsasservice/modelsasservice_controller_actions.go]: nit, typically it would be `maas-config-hash`

File: internal/controller/components/modelsasservice/modelsasservice_controller_actions.go (hunk 1 above)
Comment (ishitasequeira) [reply internal/controller/components/modelsasservice/modelsasservice_controller_actions.go:226]: I decided to keep `configMapKeyRef` with `optional: true` for backwards compatibility (pods start even without the key set), and to stay consistent with how the other env vars (`GATEWAY_NAMESPACE`, `GATEWAY_NAME`) are already wired. This is also the same pattern Kserve uses for config hash annotations. 

That said, refactoring the manifests to use direct replacements and phase out params.env could be a good future improvement to simplify this pattern. Happy to make that change as part of this PR if you think it's a better approach.

File: internal/controller/components/modelsasservice/modelsasservice_controller_actions.go [hunk 3]
```diff
@@ -201,3 +218,113 @@ func configureDestinationRule(log logr.Logger, resource *unstructured.Unstructur
...
+		return fmt.Errorf("failed to convert Deployment back to unstructured: %w", err)
+	}
+	rr.Resources[deploymentIdx].Object = u
+
+	// Also update the ConfigMap in resources (in case it was modified)
+	cmU, err := runtime.DefaultUnstructuredConverter.ToUnstructured(configMap)
+	if err != nil {
+		return fmt.Errorf("failed to convert ConfigMap back to unstructured: %w", err)
+	}
+	rr.Resources[configMapIdx].Object = cmU
```
Comment (bartoszmajsak) [inline internal/controller/components/modelsasservice/modelsasservice_controller_actions.go]: Do we need this? Wouldn't kustomize handle it through `params.env`?

**Review by bartoszmajsak — COMMENTED** [review]
Not an expert in internal operator framework, so I will leave formal review for the others. Few minor things

File: internal/controller/components/modelsasservice/modelsasservice_controller_actions.go (hunk 1 above)
Comment (bartoszmajsak) [reply internal/controller/components/modelsasservice/modelsasservice_controller_actions.go:226]: If we stick with kustomize, maybe worth trying https://github.com/kubernetes-sigs/kustomize/pull/5679 :)

File: internal/controller/components/modelsasservice/modelsasservice_controller_actions.go [hunk 4]
```diff
@@ -103,14 +114,20 @@ func customizeManifests(ctx context.Context, rr *types.ReconciliationRequest) er
...
 		"gateway-name":      maas.Spec.GatewayRef.Name,
 		"app-namespace":     appNamespace,
 	}
 
-	if err := odhdeploy.ApplyParams(rr.Manifests[0].String(), "params.env", nil, gatewayParams); err != nil {
-		return fmt.Errorf("failed to update Gateway params on path %s: %w", rr.Manifests[0].String(), err)
+	// Add API key configuration if specified
+	if maas.Spec.APIKeys != nil && maas.Spec.APIKeys.MaxExpirationDays != nil {
+		params["api-key-max-expiration-days"] = strconv.FormatInt(int64(*maas.Spec.APIKeys.MaxExpirationDays), 10)
+		log.V(4).Info("Configuring API key max expiration days", "value", *maas.Spec.APIKeys.MaxExpirationDays)
```
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply internal/controller/components/modelsasservice/modelsasservice_controller_actions.go:126]: This is validated by kubebuilder at API server level already.

File: internal/controller/components/modelsasservice/modelsasservice_controller_actions.go [hunk 5]
```diff
@@ -201,3 +218,104 @@ func configureDestinationRule(log logr.Logger, resource *unstructured.Unstructur
...
+	}
+	sort.Strings(keys)
+
+	// Build a string representation of the data
+	var sb strings.Builder
+	for _, k := range keys {
+		sb.WriteString(k)
+		sb.WriteString("=")
+		sb.WriteString(data[k])
+		sb.WriteString("\n")
```
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply internal/controller/components/modelsasservice/modelsasservice_controller_actions.go:315]: Uses length-prefixed encoding to avoid collisions when values contain special characters like '=' or newlines.

**Review by lphiri — APPROVED** [review]

File: internal/controller/components/modelsasservice/modelsasservice_controller_actions.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ugiordan) [reply internal/controller/components/modelsasservice/modelsasservice_controller_actions.go:315]: This is a valid concern.

File: internal/controller/components/modelsasservice/modelsasservice_controller_actions.go [hunk 6]
```diff
@@ -201,3 +218,104 @@ func configureDestinationRule(log logr.Logger, resource *unstructured.Unstructur
...
+		return nil
+	}
+
+	// Initialize annotations map if nil
+	if deployment.Spec.Template.Annotations == nil {
+		deployment.Spec.Template.Annotations = make(map[string]string)
+	}
+
+	// Add the config hash annotation to the pod template
+	annotationKey := labels.ODHAppPrefix + "/maas-config-hash"
```
Comment (ugiordan) [inline internal/controller/components/modelsasservice/modelsasservice_controller_actions.go:281]: This annotation key is constructed from the labels package, but the codebase has a dedicated pkg/metadata/annotations package for annotation constants (e.g. `ManagedByODHOperator`, `ManagementStateAnnotation`). Consider defining this there:

```
  // in pkg/metadata/annotations/annotations.go
  const MaaSConfigHash = "app.opendatahub.io/maas-config-hash"
```

This keeps annotation keys discoverable in one place and avoids mixing the labels package into annotation concerns.

File: internal/controller/components/modelsasservice/modelsasservice_controller_actions_test.go [hunk 7]
```diff
@@ -470,3 +474,272 @@ func createFakeClientWithoutGateway() client.Client {
...
+					},
+				},
+			}
+
+			// Verify APIKeys exists but MaxExpirationDays is nil
+			g.Expect(maas.Spec.APIKeys).ShouldNot(BeNil())
+			g.Expect(maas.Spec.APIKeys.MaxExpirationDays).Should(BeNil())
+		})
+
+		t.Run("should accept various valid maxExpirationDays values", func(t *testing.T) {
```
Comment (ugiordan) [inline internal/controller/components/modelsasservice/modelsasservice_controller_actions_test.go:544]: On the `MaxExpirationDays` field, there's a `+kubebuilder:validation:Minimum=1` but no `Maximum` ([here](https://github.com/opendatahub-io/opendatahub-operator/blob/1af82326d2a681ec7e5880c9eea3f6090cd7edd3/api/components/v1alpha1/modelsasservice_types.go#L73-L74)). Should we add a `+kubebuilder:validation:Maximum` to cap this value? Without one, the CRD accepts up to `2147483647` (the int32 max), which would effectively disable expiration.

File: internal/controller/components/modelsasservice/modelsasservice_controller_actions.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ugiordan) [reply internal/controller/components/modelsasservice/modelsasservice_controller_actions.go:315]: Even if this is not exploitable with current MaaS params (gateway names, integers), but for robustness we could add:
```
  sb.WriteString(fmt.Sprintf("%s=%q\n", k, data[k]))
```
`%q` quotes the value and escapes embedded newlines/special characters.

File: internal/controller/components/modelsasservice/modelsasservice_controller_actions_test.go [hunk 8]
```diff
@@ -470,3 +474,272 @@ func createFakeClientWithoutGateway() client.Client {
 		WithScheme(scheme).
 		Build()
 }
+
+func TestAPIKeyConfiguration(t *testing.T) {
```
Comment (ugiordan) [inline internal/controller/components/modelsasservice/modelsasservice_controller_actions_test.go:478]: These tests only verify Go struct field assignment. They construct a `ModelsAsService` and assert that its fields have the values they were just set to. They never call `customizeManifests()` and never verify that `params["api-key-max-expiration-days"]` is actually set in the rendered output.

Compare with `TestConfigureConfigHashAnnotation` below, which actually invokes the function and asserts on the result. The API key tests should follow the same pattern: call `customizeManifests()` with a prepared `ReconciliationRequest` and assert that the params map (or the resulting `params.env` content) contains the expected entry.

File: internal/controller/components/modelsasservice/modelsasservice_controller_actions.go [hunk 9]
```diff
@@ -201,3 +218,104 @@ func configureDestinationRule(log logr.Logger, resource *unstructured.Unstructur
...
+			configMap = cm
+			break
+		}
+	}
+
+	if configMap == nil {
+		log.V(1).Info("ConfigMap not found in rendered resources, skipping config hash annotation",
+			"expectedName", MaaSParametersConfigMapName)
+		return nil
+	}
```
Comment (ugiordan) [inline internal/controller/components/modelsasservice/modelsasservice_controller_actions.go:247]: This early return logs at V(1) (hidden by default) and returns nil, so if `MaaSParametersConfigMapName` or `MaaSAPIDeploymentName` are ever renamed without updating the kustomize manifests (or vice versa), this feature would silently stop working.

Consider at minimum adding a brief comment noting that these names must match the kustomize manifests, e.g.:
```
  // NOTE: MaaSParametersConfigMapName must match the ConfigMap name in the kustomize manifests.
```
